src/data/list/cycle.lean

```diff
@@ -699,4 +713,103 @@ underlying element. This representation also supports cycles that can contain du
 instance [has_repr α] : has_repr (cycle α) :=
 ⟨λ s, "c[" ++ string.intercalate ", " ((s.map repr).lists.sort (≤)).head ++ "]"⟩
 
+/-- `chain R s` means that `R` holds between adjacent elements of `s`.
+
+     chain R [a, b, c] ↔ R a b ∧ R b c ∧ R c a -/
+def chain (r : α → α → Prop) : cycle α → Prop :=
+@quotient.lift _ _ (is_rotated.setoid α) (λ l, match l with
+  | [] := (tt : Prop)
+  | (a :: m) := chain r a (m ++ [a]) end) $
+λ a b hab, begin
+  cases a with a l;
+  cases b with b m,
+  { refl },
+  { have := is_rotated_nil_iff'.1 hab,
+    contradiction },
+  { have := is_rotated_nil_iff.1 hab,
+    contradiction },
+  { change chain _ _ _ = chain _ _ _,
+    cases hab with n hn,
+    induction n with d hd generalizing a b l m,
+    { simp only [rotate_zero] at hn,
+      rw [hn.1, hn.2] },
+    { cases l with c s,
+      { simp only [rotate_singleton] at hn,
+        rw [hn.1, hn.2] },
+      { rw [nat.succ_eq_one_add, ←rotate_rotate, rotate_cons_succ, rotate_zero, cons_append] at hn,
+        rw [←hd c _ _ _ hn],
+        simp,
+        exact and.comm } } }
+end
+
+@[simp] lemma chain.nil (r : α → α → Prop) : cycle.chain r (@nil α) :=
+by exact rfl
+
+@[simp] lemma chain_cons (r : α → α → Prop) (a : α) (l : list α) :
+  chain r (a :: l) ↔ list.chain r a (l ++ [a]) :=
+iff.rfl
+
+@[simp] lemma chain_singleton (r : α → α → Prop) (a : α) : chain r [a] ↔ r a a :=
+by rw [chain_cons, nil_append, chain_singleton]
+
+lemma chain_ne_nil (r : α → α → Prop) {l : list α} (hl : l ≠ []) :
+  chain r l ↔ list.chain r (last l hl) l :=
```

Comment:
Is there a reason to prefer last over nth_le 0?

Reply:
The lemma about `nth_le 0` is quite more dubious in how useful it is, since `match` and `chain_cons` prove it immediately.

Reply:
I meant to use `list.chain r (l.nth_le 0 _) l`

Reply:
I see what you mean I think

Reply:
I don't think the lemma about `nth_le` as you wrote it is correct, `chain r [a, b]` means `r a b` and `r b a`, not `r a a` and `r a b`.

Reply:
Yeah, I see what you mean.